transforms/stage-4-new-years/index.js

```diff
@@ -0,0 +1,176 @@
+
+          return;
+        }
+
+        node.path = buildResolver(b, node.path.original, 'modifier');
+      },
+
+      ElementNode(node, nodePath) {
+        if (!(node.tag in unknownNames)) {
+          return;
+        }
+
+        if (!unknownNames[node.tag].includes('angleComponents')) {
+          console.error(
+            `In '${templatePath}', '${node.tag}' is being used as a component, but it is` +
+              ` not listed as one in the package's names map ('${namesMapPath}'). Make sure` +
+              ` stages 2 and 3 completed successfully.`
+          );
+
+          return;
+        }
+
+        if (isDefinedByLet(node, nodePath)) {
+          return;
+        }
+
+        return b.block(
+          'let',
+          [b.sexpr('component', [b.string(`${toKebabCase(node.tag)}`)])],
+          undefined,
+          b.blockItself([node], [node.tag])
+        );
+      },
+
+      MustacheStatement(node) {
+        if (!(node.path.original in unknownNames)) {
+          return;
+        }
+
+        if (!unknownNames[node.path.original].includes('ambiguousCurlies')) {
+          console.error(
+            `In '${templatePath}', '${node.path.original}' is being used as ambiguous helper or` +
+              " component, but it is not listed as one in the package's names map" +
+              ` ('${namesMapPath}'). Make sure stages 2 and 3 completed successfully.`
+          );
+
+          return;
+        }
+
+        const nameAsHelper = node.path.original + 'Helper';
+        const nameAsComponent = node.path.original + 'Component';
+        const hasParams = node.params?.length || node.hash?.pairs?.length;
+
+        const nodeAsHelper = hasParams
+          ? b.sexpr(nameAsHelper, node.params, node.hash)
+          : b.path(nameAsHelper);
+
+        const nodeAsComponent = hasParams
+          ? b.sexpr(nameAsComponent, [b.path(nameAsComponent), ...node.params], node.hash)
+          : b.path(nameAsComponent);
+
+        return b.block(
+          'let',
+          [
+            buildResolver(b, node.path.original, 'helper'),
+            buildResolver(b, node.path.original, 'component'),
+          ],
+          undefined,
+          b.blockItself(
+            [b.mustache('if', [b.path(nameAsHelper), nodeAsHelper, nodeAsComponent])],
+            [nameAsHelper, nameAsComponent]
+          )
+        );
+      },
+
+      SubExpression(node) {
+        if (!(node.path.original in unknownNames)) {
+          return;
+        }
+
+        if (!unknownNames[node.path.original].includes('helpers')) {
+          console.error(
+            `In '${templatePath}', '${node.path.original}' is being used as a helper, but it is` +
+              ` not listed as one in the package's names map ('${namesMapPath}'). Make sure` +
+              ` stages 2 and 3 completed successfully.`
+          );
+
+          return;
+        }
+
+        node.path = buildResolver(b, node.path.original, 'helper');
+      },
+    };
+  });
+};
```

Comment:
Then, here, we would want to return that list of errors, to again be processed by the caller.

(In the case that the caller is codemod-cli itself, we might need to make some tweaks in that flow, but it's still a useful pattern!)